### PR TITLE
Make slider path parsing more friendly with spans

### PR DESCRIPTION
**Pull request by huoyaoyuan**
In order to unblock #23282, this PR refactors slider parsing to not passing substrings around.

The majority of existing logic is preserved. The performance shouldn't change much with this PR, just ensuring there's no regression.

Before:

| Method                  | Mean     | Error     | StdDev    | Gen0     | Gen1     | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 1.529 ms | 0.0083 ms | 0.0065 ms | 228.5156 | 187.5000 |   4.13 MB |

After:

 | Method                  | Mean     | Error     | StdDev    | Gen0     | Gen1     | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 1.540 ms | 0.0208 ms | 0.0195 ms | 228.5156 | 187.5000 |   4.12 MB |

### CODE REVIEW

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:303]: I initially considered folding the `convertPoints` method. Now I don't want to make further change aside from unblock string parsing, it is better to separate the two conditions.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:323]: Correct. It took some time for me to figure out what `endPoint` is for.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
Comment (huoyaoyuan) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:273]: Honestly, the `buffer[count++]` pattern looks very common to me and I didn't think it's hard to read.

Comment (frenzibyte) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs]: Combined with the rest of the logic and the various kind of collections that exist in the logic (points string array, points array, segments array), it becomes very cryptic for me.

**Review by frenzibyte — APPROVED** [review]
Looks good to both my eyes and CI.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
Comment (huoyaoyuan) [inline osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:306]: Specifically, I think *this* particular case is better readable with range expression.

I don't have preference on other cases though.

Comment (frenzibyte) [reply osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:306]: There have been multiple occurrences of the `..` operator being used and us not liking it in review, I don't want that happening here so I changed it with the `Slice` method to move the PR forward.

**Comment by frenzibyte** [comment]
Requesting a second pair of eyes before proceeding with a merge, the PR is good but I'm not sure if there's something I should've checked against before approval (like triggering a diffcalc command to make sure nothing has changed with beatmap parsing or something).

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by frenzibyte** [comment]
Spreadsheets show no difference as well, so let's get this in.

**Pull request merged**
